packages/shopping-cart/src/shopping-cart-manager.ts

```diff
@@ -0,0 +1,204 @@
+	// addProductsToCart), the reducer modifies the `responseCart` stored in
+	// state. That (incomplete) data is then sent off to the server to be
+	// validated and filled-in with additional properties before being returned
+	// to the ShoppingCartManager. Because of this, optimistic updating of the
+	// cart is not possible (it may change significantly in that round trip) so
+	// we don't want to return the raw `responseCart` to the consumer. Instead,
+	// we keep a copy of the `responseCart` the last time the state had a `valid`
+	// CacheStatus and pass that to our consumers. The consumers can use
+	// `isPendingUpdate` to know when the cart data is updating.
+	const {
+		getLastValidResponseCart,
+		updateLastValidResponseCart,
+	} = createLastValidResponseCartManager( state );
+
+	const { resolveActionPromisesIfValid, addActionPromise } = createActionPromisesManager();
+	const takeActionsBasedOnState = createTakeActionsBasedOnState(
+		updateLastValidResponseCart,
+		resolveActionPromisesIfValid
+	);
+
+	// This is the main dispatcher for shopping cart actions. Dispatched actions
+	// are async and non-blocking. Most consumers can use the `subscribe`
+	// callback to know when changes have been made but if a consumer needs to
+	// know when actions are complete, they can use `dispatchAndWaitForValid`
+	// which is the dispatcher exported with the ShoppingCartManager's action
+	// creators.
+	let actionsPending = 0;
+	const dispatch = ( action: ShoppingCartAction ) => {
+		debug( `heard action request for cartKey ${ cartKey }`, action.type );
+		actionsPending++;
+		// setTimeout here defers the dispatch process until the next free cycle of
+		// the JS event loop so that dispatching actions is async and will not
+		// block any code that triggers them. Because we don't want to take certain
+		// actions if there are more pending events, we track actionsPending to be
+		// sure that all dispatched events eventually reach the reducer.
+		setTimeout( () => {
+			debug( `dispatching middleware action for cartKey ${ cartKey }`, action.type );
+			actionsPending--;
+			middleware.forEach( ( middlewareFn ) => {
+				middlewareFn( action, state, dispatch );
+			} );
+			debug( `dispatching action for cartKey ${ cartKey }`, action.type );
+			state = shoppingCartReducer( state, action );
+			takeActionsBasedOnState( state, dispatch, actionsPending > 0 );
+			notifySubscribers();
+		} );
+	};
+
+	// `dispatchAndWaitForValid` enhances the action dispatcher to return a
+	// Promise that will resolve when the cart next reaches a `valid`
+	// CacheStatus. This is the dispatcher used for all actions in the
+	// ShoppingCartManager's public API.
+	const dispatchAndWaitForValid = createDispatchAndWaitForValid( dispatch, addActionPromise );
+	const actionCreators = createActionCreators( dispatchAndWaitForValid );
+	const waitForReady = () => {
+		return new Promise< ResponseCart >( ( resolve ) => {
+			addActionPromise( resolve );
+		} );
+	};
```

Comment:
> The waitForReady method waits until the cart becomes valid, and is used mainly to wait for the initial load, is that right?

I suppose. It was added for the sake of unit tests. React code should never need it because of `subscribe` and action Promises.

> Therefore, I propose that createShoppingCartManager merely allocates the object and does nothing else, and that the initial fetch is done explicitly with a new method.

That's a good idea. I'll try it out.

Reply:
I wonder what should happen if you try to take cart actions before calling `fetchInitialCart` 🤔 Should there be a warning? A rejected promise? Or should we let the action silently queue, waiting until the consumer calls the fetch?

Reply:
Either the action could just silently queue, or it could throw a rejected promise. The condition that throws an error seems quite simple: never allow any update, not even being queued, if the `cacheStatus` is `fresh`. We need to at least advance from `fresh` to `fresh-pending` by calling `fetchInitialCart` before allowing any update to be performed or even queued.

In practice the cart will be used mostly through the React bindings, which never forget to call `fetchInitialCart`, so I think this is not a big issue.

Reply:
It turns out that there's sort of a hidden feature! If you dispatch an action to modify the cart and it hasn't been fetched, that action will automatically be queued and the initial cart will be fetched first. This is because we already queue actions if the cart is `fresh`, and when the state-based actions run after that action is queued, they notice that the cart is `fresh` and dispatch the fetch action.

I actually like this a lot because if a client (without using the hook) wants to add something to the cart this is one less step for them to do and I can't think of any case where someone would want to _not_ do it.